flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1527 @@
+		ByteBuffer targetKeyByteBuffer = chunk.getByteBuffer(offsetInChunk);
+		int offsetInByteBuffer = chunk.getOffsetInByteBuffer(offsetInChunk);
+
+		int level = SkipListUtils.getLevel(targetKeyByteBuffer, offsetInByteBuffer);
+		int targetKeyOffset = offsetInByteBuffer + SkipListUtils.getKeyDataOffset(level);
+
+		return SkipListKeyComparator.compareNamespaceAndNode(namespaceByteBuffer, namespaceOffset, namespaceLen,
+			targetKeyByteBuffer, targetKeyOffset);
+	}
+
+	/**
+	 * Update the value of the node with copy-on-write mode. The old value will
+	 * be linked after the new value, and can be still accessed.
+	 *
+	 * @param node  the node to update.
+	 * @param value the value.
+	 * @return the old value pointer.
+	 */
+	private long updateValueWithCopyOnWrite(long node, byte[] value) {
+		// a null value indicates this is a removed node
+		int valueSize = value == null ? 0 : value.length;
+		int totalValueLen = SkipListUtils.getValueMetaLen() + valueSize;
+		long valuePointer = spaceAllocator.allocate(totalValueLen);
+
+		Chunk nodeChunk = spaceAllocator.getChunkById(SpaceUtils.getChunkIdByAddress(node));
+		int offsetInNodeChunk = SpaceUtils.getChunkOffsetByAddress(node);
+		ByteBuffer nodeByteBuffer = nodeChunk.getByteBuffer(offsetInNodeChunk);
+		int offsetInNodeByteBuffer = nodeChunk.getOffsetInByteBuffer(offsetInNodeChunk);
+		long oldValuePointer = SkipListUtils.getValuePointer(nodeByteBuffer, offsetInNodeByteBuffer);
+
+		doWriteValue(valuePointer, value, stateMapVersion, node, oldValuePointer);
+
+		// update value pointer in node after the new value has points the older value so that
+		// old value can be accessed concurrently
+		SkipListUtils.putValuePointer(nodeByteBuffer, offsetInNodeByteBuffer, valuePointer);
+
+		return oldValuePointer;
+	}
+
+	/**
+	 * Update the value of the node with replace mode. The old value will be unlinked and replaced
+	 * by the new value, and can not be accessed later. Note that the space of the old value
+	 * is not freed here, and the caller of this method should be responsible for the space management.
+	 *
+	 * @param node  the node whose value will be replaced.
+	 * @param value the value.
+	 * @return the old value pointer.
+	 */
+	private long updateValueWithReplace(long node, byte[] value) {
+		// a null value indicates this is a removed node
+		int valueSize = value == null ? 0 : value.length;
+		int totalValueLen = SkipListUtils.getValueMetaLen() + valueSize;
+		long valuePointer = spaceAllocator.allocate(totalValueLen);
+
+		Chunk nodeChunk = spaceAllocator.getChunkById(SpaceUtils.getChunkIdByAddress(node));
+		int offsetInNodeChunk = SpaceUtils.getChunkOffsetByAddress(node);
+		ByteBuffer nodeByteBuffer = nodeChunk.getByteBuffer(offsetInNodeChunk);
+		int offsetInNodeByteBuffer = nodeChunk.getOffsetInByteBuffer(offsetInNodeChunk);
+
+		long oldValuePointer = SkipListUtils.getValuePointer(nodeByteBuffer, offsetInNodeByteBuffer);
```

Comment:
Because the OO approach will produce a lot of objects which will affect GC a lot. Please refer to [the analysis of JDK CSLM implementation](https://docs.google.com/document/d/16VIY7o-18sM-pIlIYkbTuhKPmwfnqabCt_nlOARAzdg/edit#) and a compacted data structure we introduced for HBase to reduce GC pressure. Search for `Key space schema` and `Value space schema` in `SkipListUtils` and we could find a similar design here.

Reply:
Would it be possible to reuse a single object? One could have a decorator which one initializes with the current pointer one is working on and which offers all the methods to access the fields as we currently have in `SkipListUtils`. This should not add too much pressure on the GC.

I couldn't find any numbers in the linked document (only skimmed over it). Could you provide me with the concrete paragraph where I can find the performance comparison?

Reply:
Unified replies/discussions in other comments, so close the conversation here.